input/otlp/logs_test.go

```diff
@@ -225,34 +226,43 @@ func TestConsumerConsumeLogs(t *testing.T) {
 
 func TestConsumeLogsSemaphore(t *testing.T) {
 	logs := plog.NewLogs()
-	var batches []*modelpb.Batch
 
+	var once sync.Once
```

Comment:
I think, we don't really need this complicated concurrency primitive structure here and can simplify this. The same test can be accomplished with a single buffered channel without any additional primitives, see the example below.

```
func TestConsumeLogsSemaphore(t *testing.T) {
	logs := plog.NewLogs()
	var batches []*modelpb.Batch

	doneCh := make(chan struct{})
	recorder := modelpb.ProcessBatchFunc(func(ctx context.Context, batch *modelpb.Batch) error {
		doneCh <- struct{}{}
		doneCh <- struct{}{}
		batchCopy := batch.Clone()
		batches = append(batches, &batchCopy)
		return nil
	})
	consumer := otlp.NewConsumer(otlp.ConsumerConfig{
		Processor: recorder,
		Semaphore: semaphore.NewWeighted(1),
	})

	go func() {
		_, err := consumer.ConsumeLogsWithResult(context.Background(), logs)
		assert.NoError(t, err)
	}()

	<-doneCh
	ctx, cancel := context.WithTimeout(context.Background(), time.Millisecond)
	defer cancel()
	_, err := consumer.ConsumeLogsWithResult(ctx, logs)
	assert.Equal(t, err.Error(), "context deadline exceeded")
	<-doneCh

	// turn channel into sink
	doneCh = make(chan struct{}, 2)
	_, err = consumer.ConsumeLogsWithResult(context.Background(), logs)
	assert.NoError(t, err)
}
```